File: src-tauri/src/domain/template.rs
use serde::{Deserialize, Serialize};

use super::field::FieldDef;
use super::layout::LayoutSection;

/// Mirror of the TS `Template` type.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Template {
    pub id: String,
    pub name: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub parent_id: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub icon: Option<String>,
    pub fields: Vec<FieldDef>,
    /// Optional infobox layout. Fields not referenced by any layout
    /// item are rendered after the layout, in declaration order, so
    /// adding a field never makes it disappear.
    #[serde(default)]
    pub layout: Vec<LayoutSection>,
}

File: src/types/template.ts
import type { FieldDef } from "./field";
import type { LayoutSection } from "./layout";

/**
 * A user-editable template that defines the infobox fields for an entry type.
 * Templates live as JSON files inside `<project>/templates/<id>.json`.
 *
 * Common fields (name, summary, tags, body, images) are not declared here:
 * they are intrinsic to every entry regardless of template.
 */
export type Template = {
  id: string;
  name: string;
  /** Optional parent template; sub-templates inherit and extend its fields. */
  parentId?: string;
  /** Emoji or icon identifier displayed next to entries of this template. */
  icon?: string;
  fields: FieldDef[];
  /** Optional infobox layout. Fields not referenced fall through to a
   * trailing block so adding a field never makes it disappear. */
  layout: LayoutSection[];
};
